validation: improve performance of CheckBlockIndex

mzumsande opened the pull request:
`CheckBlockIndex() ` are consistency checks that are currently enabled by default on regtest.

The function is rather slow, which is annoying if you
* attempt to run it on other networks, especially if not fully synced
* want to generate a long chain on regtest and see block generation slow down because you forgot to disable `-checkblockindex` or don't know it existed.

One reason why it's slow is that in order to be able to traverse the block tree depth-first from genesis, it inserts pointers to all block indices into a `std::multimap` - for which inserts and lookups become slow once there are hundred thousands of entries.
However, typically the block index is mostly chain-like with just a few forks so a multimap isn't really needed for the most part. This PR suggests to store the block indices of the chain ending in the best header in a vector instead, and store only the rest of the indices in a multimap. This does not change the actual consistency checks that are being performed for each index, just the way the block index tree is stored and traversed.

This adds a bit of complication to make sure each block is visited (note that there are asserts that check it), making sure that the two containers are traversed correctly, but it speeds up the function considerably:

On master, a single invocation of `CheckBlockIndex` takes ~1.4s on mainnet for me (4.9s on testnet which has >2.4 million blocks).
With this branch, the runtime goes down to ~0.27s (0.85s on testnet).This is a speedup by a factor ~5.

mzumsande committed:
validation: improve performance of CheckBlockIndex

by not saving all indexes in a std::multimap, but only
those that are not part of the best header chain.
The indexes of the best header chain are stored in a vector,
which, in the typical case of a mostly linear chain with
a few forks, results in a much smaller multimap, and increases
performance noticeably for long chains.

This does not change the actual consistency checks that are being
performed for each index, just the way the block index tree is
stored and traversed.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

fanquake commented:
cc @martinus

martinus commented:
Hi, I wonder how you measured the runtime, is there a benchmark? Another optimization that might be worth it, it is to replace the `[multimap](std::multimap<CBlockIndex*,CBlockIndex*> forward)` with an `std::unordered_map<CBlockIndex*, std::vector<CBlockIndex*>> forward`.

mzumsande commented on src/validation.cpp:
```diff
@@ -4698,18 +4698,31 @@ void ChainstateManager::CheckBlockIndex()
+    // All remaining blocks are stored in a multimap.
+    CBlockIndex* index_walk{m_best_header};
+    const int best_hdr_height{m_best_header->nHeight};
+    std::vector<CBlockIndex*> best_hdr_chain(best_hdr_height + 1);
```
I didn't do that so that the speedup is independent from the chain, and also applies to situations where we have all the headers, but `m_chain` is behind (IBD, reindex). Which seemed important because these are the exact situations where `CheckBlockIndex` will be called a lot and slows `bitcoind` down most noticeably.

Sjors commented:
While you're at it, I noticed that the help text for `-checkblockindex` says "occasionally" (added in #7290), but afaik it happens every block (and every header during the initial header sync). We should probably drop that word, or alternatively provide an `interval` argument.

mzumsande commented:
> While you're at it, I noticed that the help text for -checkblockindex says "occasionally" (added in https://github.com/bitcoin/bitcoin/pull/7290), but afaik it happens every block (and every header during the initial header sync). We should probably drop that word, or alternatively provide an interval argument.

Good point, I think that an interval argument similar to addrman or mempool consistency checks makes sense. Will try that in addition to the other points above (it might take a few weeks though, therefore put the PR into draft).

mzumsande committed:
bench: add benchmark for checkblockindex

mzumsande committed:
validation: allow to specify frequency for -checkblockindex

This makes it similar to -checkaddrman and -checkmempool, which
also allow to run the check occasionally instead of always / never.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

mzumsande commented:
I added a commit with a benchmark, and another one that allow to specify the frequency for `-checkblockindex` (analogous to the way `-checkaddrman` and `-checkmempool` work).

> Another optimization that might be worth it, it is to replace the std::multimap<CBlockIndex*,CBlockIndex*> forward with an std::unordered_map<CBlockIndex*, std::vector<CBlockIndex*>> forward.

I tried that in `https://github.com/mzumsande/bitcoin/commit/51aae54cfafc9700581e69dc129866cff76925ae` (directly on master without the changes of this PR), however it doesn't seem to improve performance - I'll post benchmark results below.

mzumsande commented:
Benchmark results:
<details>
<summary>master</summary>

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          331,165.67 |            3,019.64 |    0.3% |      0.01 | `CheckBlockIndex`
|          320,603.00 |            3,119.12 |    1.7% |      0.01 | `CheckBlockIndex`
|          315,683.67 |            3,167.73 |    0.4% |      0.01 | `CheckBlockIndex`
|          329,077.33 |            3,038.80 |    0.3% |      0.01 | `CheckBlockIndex`
|          315,345.33 |            3,171.13 |    0.4% |      0.01 | `CheckBlockIndex`

</details>

<details>
<summary>PR</summary>

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          127,273.00 |            7,857.13 |    1.8% |      0.01 | `CheckBlockIndex`
|          123,687.22 |            8,084.91 |    0.1% |      0.01 | `CheckBlockIndex`
|          136,637.17 |            7,318.65 |    0.1% |      0.01 | `CheckBlockIndex`
|          123,805.67 |            8,077.17 |    0.3% |      0.01 | `CheckBlockIndex`
|          124,162.62 |            8,053.95 |    0.2% |      0.01 | `CheckBlockIndex`
</details>

<details>
<summary>martinus' idea</summary>

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          384,395.00 |            2,601.49 |    0.5% |      0.01 | `CheckBlockIndex`
|          380,674.67 |            2,626.92 |    0.3% |      0.01 | `CheckBlockIndex`
|          384,337.33 |            2,601.88 |    1.1% |      0.01 | `CheckBlockIndex`
|          410,836.50 |            2,434.06 |    3.4% |      0.01 | `CheckBlockIndex`
|          384,461.33 |            2,601.04 |    2.0% |      0.01 | `CheckBlockIndex`
</details>

The benchmark currently uses a linear chain of 1100 headers. In general, the efficiency gain becomes larger with a longer chain, for example with just 100 headers the speedup is less.

So there is a speedup of a facto >2.5 with the benchmark (that becomes larger on larger chains, see results for mainnet / testnet in OP).
The suggestion by @martinus to use `std::unordered_map<CBlockIndex*, std::vector<CBlockIndex*>>` doesn't seem to improve performance (though it's also very possible that I might have not implemented it in a great away).

mzumsande commented:
[87d6155 ](https://github.com/bitcoin/bitcoin/commit/87d6155b91797d795433c9c8bca5da20c74f810c)to [efdc8e3](https://github.com/bitcoin/bitcoin/commit/efdc8e3a0ee81b143ee17a578f1e27b1c427a912): rebased

mzumsande commented on src/validation.cpp:
```diff
@@ -4817,6 +4817,13 @@ void ChainstateManager::LoadExternalBlockFile(
+bool ChainstateManager::ShouldCheckBlockIndex() const
+{
+    if (!*Assert(m_options.check_block_index)) return false;
+    if (GetRand(*m_options.check_block_index) >= 1) return false;
```
I think it would be nice to have consistency among `-checkaddrman`, `-checkmempool` and `-checkblockindex`. Since non-determinism would only occur for values other than 0 and 1, maybe it'd be best not to use other values than that in tests? Or change all three do be deterministic?

furszy reviewed (commented):
Concept ACK.
Will start reviewing next week.

furszy commented on src/validation.cpp:
```diff
@@ -4833,18 +4833,31 @@ void ChainstateManager::CheckBlockIndex()
+    // Build forward-pointing data structure for the entire block tree.
+    // For performance reasons, indexes of the best header chain are stored in a vector.
+    // All remaining blocks are stored in a multimap.
+    CBlockIndex* index_walk{m_best_header};
```
Haven't gone deeper yet but this reminds me to #26245 (and #26260).

mzumsande commented on src/bench/checkblockindex.cpp:
```diff
@@ -0,0 +1,20 @@
+
+static void CheckBlockIndex(benchmark::Bench& bench)
+{
+    auto testing_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    // Mine some more blocks
+    testing_setup->mineBlocks(1000);
+    bench.run([&] {
+        testing_setup->m_node.chainman->CheckBlockIndex();
+    });
```
no, I hadn't thought about it, but that sounds like a good idea (which seems independent from this PR though)

ryanofsky commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -24,7 +24,7 @@
 util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManager::Options& opts)
 {
-    if (auto value{args.GetBoolArg("-checkblockindex")}) opts.check_block_index = *value;
+    if (auto value{args.GetIntArg("-checkblockindex")}) opts.check_block_index = *value;
```
In commit "validation: allow to specify frequency for -checkblockindex" (efdc8e3a0ee81b143ee17a578f1e27b1c427a912)

A problem with this change is that it will now interpret bare `-checkblockindex` arguments as `-checkblockindex=0` instead of `-checkblockindex=1`. Could fix this with:

```c++
if (auto value{args.GetIntArg("-checkblockindex")}) {
    // Interpret bare -checkblockindex argument as 1 instead of 0.
   opts.check_block_index = args.GetArg("-checkblockindex")->empty() ? 1 : *value;
}
```

mzumsande replied:
Good catch! Took your suggestion. I copied this from the `-checkmempool` parsing code, which will still interpret a bare `-checkmempool` as `-checkmempool=0` (even though that was a boolean arg originally as well). Maybe that should be changed as well.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4833,18 +4833,31 @@ void ChainstateManager::CheckBlockIndex()
+    // Build forward-pointing data structure for the entire block tree.
+    // For performance reasons, indexes of the best header chain are stored in a vector.
+    // All remaining blocks are stored in a multimap.
+    CBlockIndex* index_walk{m_best_header};
+    const int best_hdr_height{m_best_header->nHeight};
+    std::vector<CBlockIndex*> best_hdr_chain(best_hdr_height + 1);
+    while (index_walk) {
+        best_hdr_chain[index_walk->nHeight] = index_walk;
+        index_walk = index_walk->pprev;
+    }
```
In commit "validation: improve performance of CheckBlockIndex" (0f7963e0b306257ae5d547ba979712156c9d3a30)

This seems like this is reimplementing what's already implemented in the CChain class, and all this initialization could be replaced with:

```c++
CChain best_headers;
if (m_best_header) best_headers.SetTip(m_best_header});
```

Also code below could be simplified.

mzumsande replied:
nice! Changed it to use `CChain`.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4833,18 +4833,31 @@ void ChainstateManager::CheckBlockIndex()
     for (auto& [_, block_index] : m_blockman.m_block_index) {
-        forward.emplace(block_index.pprev, &block_index);
+        // Only save indexes in forward that are not part of the best header chain.
+        if (block_index.nHeight > best_hdr_height || best_hdr_chain[block_index.nHeight]->GetBlockHash() != block_index.GetBlockHash()) {
```
In commit "validation: improve performance of CheckBlockIndex" (0f7963e0b306257ae5d547ba979712156c9d3a30)

Calling GetBlockHash() here is unnecessary, right? It seems like it would be more consistent with rest of code to just compare pointers not hashes.

mzumsande replied:
Makes sense, but since I replaced then entire condition with `CChain::Contains()`, this is no longer relevant.

ryanofsky reviewed (commented):
Code review efdc8e3a0ee81b143ee17a578f1e27b1c427a912.

Concept ACK. Seems like this gives a significant speedup from a simple code change.

I left some questions below and didn't ack yet because I think the current implementation might inadvertently break passing a `-checkblockindex` argument with no value, but otherwise this looks good.

mzumsande commented:
just rebased, will address feedback soon (I have a local merge conflict which apparently doesn't bother Github)

mzumsande commented:
[08a9b35 ](https://github.com/bitcoin/bitcoin/commit/08a9b356dc99ff5d7f1d64981848dee9d22840c4)to [4f51a4c](https://github.com/bitcoin/bitcoin/commit/4f51a4c01f2ce8ee1390e3b7a373d91158d8a83e): Addressed feedback by @ryanofsky, thanks!

furszy reviewed (commented):
Code review ACK 4f51a4c

Would be nice to cover this function properly with a good number of unit tests. It shouldn't be hard, the `validation_block_tests.cpp` file contain all the util functions to create different chain topologies.

mzumsande commented on src/validation.cpp:
```diff
@@ -5053,20 +5075,34 @@ void ChainstateManager::CheckBlockIndex()
                 pindex = pindexPar;
                 nHeight--;
                 continue;
             }
```
done!

mzumsande commented:
[4f51a4c ](https://github.com/bitcoin/bitcoin/commit/4f51a4c01f2ce8ee1390e3b7a373d91158d8a83e)to [7fcd686](https://github.com/bitcoin/bitcoin/commit/7fcd686f1fe219d8c4cb4ba90127d1139eee9d22): Rebased, addrressed nit.

> Would be nice to cover this function properly with a good number of unit tests. It shouldn't be hard, the `validation_block_tests.cpp` file contain all the util functions to create different chain topologies.

I agree, might take me a few weeks to get to that thoughs since I have a few other things on my list.
Although just mentioning that quite a few constellations are already covered in random unit / functional tests that are targeted at something else - when I implemented this originally and didn't get some details right, I got several failures.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5034,6 +5034,13 @@ void ChainstateManager::LoadExternalBlockFile(
 
+bool ChainstateManager::ShouldCheckBlockIndex() const
+{
+    if (!*Assert(m_options.check_block_index)) return false;
```
In commit "validation: allow to specify frequency for -checkblockindex" (b51eb422d657bb1d0e6f9c7e71eb8a87a8c74f42)

Note: I was initially confused why this is has an Assert to check if the option is set, instead of just returning a default value when the option is unset, but the reason seems to be that this is a check to make sure that Flatten() has been called and `opts.chainparams.DefaultConsistencyChecks()` has been applied. Could maybe add a comment above this like "// Assert to verify Flatten() has been called." to clarify this

mzumsande replied:
I added a comment. By the way, the Assert is also already present on master.

ryanofsky reviewed (approved):
Code review ACK b51eb422d657bb1d0e6f9c7e71eb8a87a8c74f42 but it seems like code could be simpler and clearer (could eliminate finished and in_best_hdr_chain variables and redundant height checks before using CChain::operator[])

mzumsande commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -24,8 +24,10 @@
 util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManager::Options& opts)
 {
-    if (auto value{args.GetBoolArg("-checkblockindex")}) opts.check_block_index = *value;
-
```
done

mzumsande commented on src/validation.cpp:
```diff
@@ -5050,18 +5050,26 @@ void ChainstateManager::CheckBlockIndex()
-    assert(rangeGenesis.first == rangeGenesis.second); // There is only one index entry with parent nullptr.
+    CBlockIndex* pindex = best_hdr_chain[0];
+    assert(pindex);
+    bool in_best_hdr_chain{true};
```
I changed it as suggested, and didn't see a noticeable impact on performance.

mzumsande commented on src/validation.cpp:
```diff
@@ -5307,20 +5325,32 @@ void ChainstateManager::CheckBlockIndex()
+                // The parent is part of the best header chain, all forks have been processed.
+                in_best_hdr_chain = true;
+                if (nHeight <= m_best_header->nHeight) {
+                    // If possible, continue by processing a sibling of the best header chain.
```
fixed

mzumsande commented on src/validation.cpp:
```diff
@@ -5274,17 +5282,27 @@ void ChainstateManager::CheckBlockIndex()
+        if (in_best_hdr_chain) break; // we are finished, since the best header chain is alway processed last
+
         // Move upwards until we reach a node of which we have not yet visited the last child.
+        bool finished{false};
```
Done.

mzumsande commented on src/validation.cpp:
```diff
@@ -5307,20 +5325,32 @@ void ChainstateManager::CheckBlockIndex()
-                pindex = pindexPar;
-                nHeight--;
-                continue;
+            } else if (pindexPar->nHeight <= best_hdr_chain.Height() && pindexPar == best_hdr_chain[nHeight - 1]) {
```
True, changed.

mzumsande commented on src/validation.cpp:
```diff
@@ -5307,20 +5325,32 @@ void ChainstateManager::CheckBlockIndex()
             }
+            // Move up further.
+            pindex = pindexPar;
+            nHeight--;
```
Nice! I took that suggestion.

mzumsande commented:
[b51eb42](https://github.com/bitcoin/bitcoin/commit/b51eb422d657bb1d0e6f9c7e71eb8a87a8c74f42) to [e880ee0](https://github.com/bitcoin/bitcoin/commit/e880ee0a634f0be5079d39328fa5e118452553a9)
Addressed feedback by @ryanofsky - now the diff to master should be a bit smaller than before.

Still haven't had the time to write more unit tests for `CheckBlockIndex()`.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5050,19 +5050,25 @@ void ChainstateManager::CheckBlockIndex()
+    // For performance reasons, indexes of the best header chain are stored in a vector (within CChain).
+    // All remaining blocks are stored in a multimap.
+    CChain best_hdr_chain;
+    if (m_best_header) best_hdr_chain.SetTip(*m_best_header);
```
In commit "validation: improve performance of CheckBlockIndex" (4b05c654c1a5df294fa32a94c38001ac2b4ff2de)

It seems like this code cannot work in the case where `m_best_header` is null because `assert(pindex)` on line 5071 below will be triggered.

I think it might make sense to replace `if (m_best_header)` here with `assert(m_best_header);` to make the assumption explicit. Otherwise, I think you will need to set the `best_hdr_chain` tip to something else, like `blockman.m_block_index.begin()`

mzumsande replied:
changed to an assert.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5307,9 +5321,14 @@ void ChainstateManager::CheckBlockIndex()
+            } else if (pindexPar == best_hdr_chain[nHeight - 1]) {
+                // Move to pindex's sibling on the best-chain, if it has one.
+                pindex = best_hdr_chain[nHeight];
+                assert(pindex || pindexPar == m_best_header);
```
In commit "validation: improve performance of CheckBlockIndex" (4b05c654c1a5df294fa32a94c38001ac2b4ff2de)

I think this assert might be a little clearer if it replaced `m_best_header` with `best_hdr_chain.Tip()`. Both values should be identical but latter doesn't require remembering how the `best_hdr_chain` tip was assigned ~300 lines earlier in the code. Also I think the check could be stricter because if `pindex` is not null, `pindexPar` cannot be the best header. Would suggest:

```c++
// There will not be a next block if (and only if) parent block is the best header.
assert((pindex == nullptr) == (pindexPar == best_hdr_chain.Tip()));
```

mzumsande replied:
done, thanks!

ryanofsky reviewed (approved):
Code review ACK e880ee0a634f0be5079d39328fa5e118452553a9. Just rebase and suggested simplifications since the last review

furszy commented on src/validation.cpp:
```diff
@@ -5050,19 +5050,25 @@ void ChainstateManager::CheckBlockIndex()
+    // For performance reasons, indexes of the best header chain are stored in a vector (within CChain).
+    // All remaining blocks are stored in a multimap.
+    CChain best_hdr_chain;
+    if (m_best_header) best_hdr_chain.SetTip(*m_best_header);
```
In 4b05c654c1:

Using `m_best_header` shouldn't be working (at least not as you expect) for `invalidateblock`.

During `InvalidateBlock()`, we first disconnect all blocks, then call `CheckBlockIndex` without updating `m_best_header` (this is done later on the flow, when we call `InvalidChainFound()`). So, by the time `CheckBlockIndex` is called, `best_hdr_chain` is loaded with a chain that is not the active one.

Any reason not to use `ActiveTip()` instead of `m_best_header` here?

ryanofsky replied:
> Any reason not to use `ActiveTip()` instead of `m_best_header` here?

I think the reason is just that we want `best_hdr_chain` to be the longest chain with the most forks for the optimization to work well. We don't actually care which chain it is because CheckBlockIndex will traverse all the blocks in any case. If a different chain is chosen it just means the blocks will be traversed in a different order, less or more efficiently.

furszy replied:
ok, yeah. Then it is not an issue but, isn't error-prone to call `best_hdr_chain` something that might not be the best chain then? E.g. it seems that if we ever add a check only for the active chain blocks here, the `best_hdr_chain.Contains()` usage temptation will be there.

ryanofsky replied:
> ok, yeah. Then it is not an issue but, isn't error-prone to call `best_hdr_chain` something that might not be the best chain then? E.g. it seems that if we ever add a check only for the active chain blocks here, the `best_hdr_chain.Contains()` usage temptation will be there.

I don't love the name `best_hdr_chain` but don't think it is too bad. I agree it could potentially be confused with the active chain in new code, and that would be a bug. In my original suggestion https://github.com/bitcoin/bitcoin/pull/28339#discussion_r1522152244, I called it `best_headers` so at least the name was consistent with `m_best_header`. I don't really have other ideas though.

mzumsande replied:
> Any reason not to use ActiveTip() instead of m_best_header here

My main reason is that `CheckBlockIndex()` being independent from the active chainstate makes the performance improvement applicable to situations where we have all the headers but haven't validated much of the chain: See  [this old answer](https://github.com/bitcoin/bitcoin/pull/28339#discussion_r1307561848). IBD and reindex seem much more relevant than extreme `invalidateblock` scenarios.

mzumsande replied:
> I agree it could potentially be confused with the active chain in new code, and that would be a bug.

I added a comment in `CheckBlockIndex()` describing why these can differ.

furszy commented:
@DrahtBot, I just reviewed it..

furszy commented on src/validation.cpp:
```diff
@@ -5050,19 +5050,25 @@ void ChainstateManager::CheckBlockIndex()
-        forward.emplace(block_index.pprev, &block_index);
+        // Only save indexes in forward that are not part of the best header chain.
+        if (!best_hdr_chain.Contains(&block_index)) {
+            // Only genesis, which must be be part of the best header chain, can have a nullptr parent.
```
tiny typo
```suggestion
            // Only genesis, which must be part of the best header chain, can have a nullptr parent.
```

mzumsande replied:
fixed

mzumsande commented:
[e880ee0 ](https://github.com/bitcoin/bitcoin/commit/e880ee0a634f0be5079d39328fa5e118452553a9)to [3c3895c](https://github.com/bitcoin/bitcoin/commit/3c3895ccfa108e10f4a0b1a78f64ffedade1fd11):
Addressed review feedback.

ryanofsky reviewed (approved):
Code review ACK 3c3895ccfa108e10f4a0b1a78f64ffedade1fd11

Left another suggestion, but this looks good in its current form.

mzumsande commented on src/validation.cpp:
```diff
@@ -5274,16 +5283,24 @@ void ChainstateManager::CheckBlockIndex()
+            continue;
         }
         // This is a leaf node.
+        if (best_hdr_chain.Contains(pindex)) break; // we are finished, since the best header chain is alway processed last
```
done as suggested.

mzumsande commented:
[3c3895c ](https://github.com/bitcoin/bitcoin/commit/3c3895ccfa108e10f4a0b1a78f64ffedade1fd11)to [5bc2077](https://github.com/bitcoin/bitcoin/commit/5bc2077e8f592442b089affdf0b5795fbc053bb8):
Addressed review feedback

furszy reviewed (commented):
ACK 5bc2077e8f592442b089affdf0b5795fbc053bb8

ryanofsky reviewed (approved):
Code review ACK 5bc2077e8f592442b089affdf0b5795fbc053bb8. Just added suggested assert and simplification since last review

achow101 commented:
ACK 5bc2077e8f592442b089affdf0b5795fbc053bb8

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/290.
